src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -170,49 +170,68 @@ internal static BoundStatement BindUsingStatementOrDeclarationFromParts(SyntaxNo
                 iDisposableConversion = classifyConversion(fromExpression, disposableInterface, ref useSiteDiagnostics);
-                disposeMethodOpt = null;
-                awaitableTypeOpt = null;
+                patternDisposeInfo = null;
+                awaitableType = null;
 
                 diagnostics.Add(syntax, useSiteDiagnostics);
 
                 if (iDisposableConversion.IsImplicit)
                 {
                     if (hasAwait)
                     {
-                        awaitableTypeOpt = originalBinder.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_ValueTask);
+                        awaitableType = originalBinder.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_ValueTask);
                     }
                     return true;
                 }
 
-                TypeSymbol type = fromExpression ? expressionOpt.Type : declarationTypeOpt;
+                Debug.Assert(!fromExpression || expressionOpt != null);
+                TypeSymbol? type = fromExpression ? expressionOpt!.Type : declarationTypeOpt;
 
                 // If this is a ref struct, or we're in a valid asynchronous using, try binding via pattern.
                 // We won't need to try and bind a second time if it fails, as async dispose can't be pattern based (ref structs are not allowed in async methods)
                 if (type is object && (type.IsRefLikeType || hasAwait))
                 {
-                    BoundExpression receiver = fromExpression
+                    BoundExpression? receiver = fromExpression
                                                ? expressionOpt
                                                : new BoundLocal(syntax, declarationsOpt[0].LocalSymbol, null, type) { WasCompilerGenerated = true };
 
                     DiagnosticBag patternDiagnostics = originalBinder.Compilation.IsFeatureEnabled(MessageID.IDS_FeatureUsingDeclarations)
                                                        ? diagnostics
                                                        : new DiagnosticBag();
-                    disposeMethodOpt = originalBinder.TryFindDisposePatternMethod(receiver, syntax, hasAwait, patternDiagnostics);
-                    if (disposeMethodOpt is object)
+                    MethodSymbol disposeMethod = originalBinder.TryFindDisposePatternMethod(receiver, syntax, hasAwait, patternDiagnostics);
+                    if (disposeMethod is object)
                     {
                         MessageID.IDS_FeatureUsingDeclarations.CheckFeatureAvailability(diagnostics, originalBinder.Compilation, syntax.Location);
+
+                        var argumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(disposeMethod.ParameterCount);
+                        ImmutableArray<int> argsToParams = default;
+                        bool expanded = disposeMethod.HasParamsParameter();
+                        originalBinder.BindDefaultArguments(
```

Comment:
Did we already have tests for using declaration where the Dispose/Async method has default parameter values?

Reply:
Non-async does not match with params/default. We have `await using ()` tests with a default parameter (`CodeGenAwaitUsingTests.TestPatternBasedDisposal_InstanceMethod_ReturningInt`), but not using declaration tests, so I'll add one.

---
In reply to: [549493661](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493661) [](ancestors = 549493661)